sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala

```diff
@@ -85,22 +85,48 @@ abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
       case ReturnAnswer(rootPlan) => rootPlan match {
         case Limit(IntegerLiteral(limit), Sort(order, true, child))
             if limit < conf.topKSortFallbackThreshold =>
-          TakeOrderedAndProjectExec(limit, order, child.output, planLater(child)) :: Nil
+          TakeOrderedAndProjectExec(limit, 0, order, child.output, planLater(child)) :: Nil
         case Limit(IntegerLiteral(limit), Project(projectList, Sort(order, true, child)))
             if limit < conf.topKSortFallbackThreshold =>
-          TakeOrderedAndProjectExec(limit, order, projectList, planLater(child)) :: Nil
+          TakeOrderedAndProjectExec(limit, 0, order, projectList, planLater(child)) :: Nil
         case Limit(IntegerLiteral(limit), child) =>
-          CollectLimitExec(limit, planLater(child)) :: Nil
+          CollectLimitExec(limit, 0, planLater(child)) :: Nil
+        case GlobalLimitAndOffset(IntegerLiteral(limit), IntegerLiteral(offset),
+          Sort(order, true, child)) if limit + offset < conf.topKSortFallbackThreshold =>
+          TakeOrderedAndProjectExec(limit, offset, order, child.output, planLater(child)) :: Nil
+        case GlobalLimitAndOffset(IntegerLiteral(limit), IntegerLiteral(offset),
+          Project(projectList, Sort(order, true, child)))
+            if limit + offset < conf.topKSortFallbackThreshold =>
+          TakeOrderedAndProjectExec(limit, offset, order, projectList, planLater(child)) :: Nil
+        case GlobalLimitAndOffset(IntegerLiteral(limit), IntegerLiteral(offset), child) =>
+          CollectLimitExec(limit, offset, planLater(child)) :: Nil
         case Tail(IntegerLiteral(limit), child) =>
           CollectTailExec(limit, planLater(child)) :: Nil
         case other => planLater(other) :: Nil
       }
       case Limit(IntegerLiteral(limit), Sort(order, true, child))
           if limit < conf.topKSortFallbackThreshold =>
-        TakeOrderedAndProjectExec(limit, order, child.output, planLater(child)) :: Nil
+        TakeOrderedAndProjectExec(limit, 0, order, child.output, planLater(child)) :: Nil
```

Comment:
It seems like there is some duplication between this and the cases under the `ReturnAnswer` block in L86-L105. Could we deduplicate them int once place? It could help us make sure the numbers are correct when adding the limit and offset together and comparing against the `conf.topKSortFallbackThreshold`.

Reply:
Maybe. But let makes this PR more pure.

Reply:
Could we finish it in another PR ?

Reply:
Sure, it is OK to leave a TODO comment for this.

Reply:
I found the code is different between the two places.